### PR TITLE
go-modules: don't set trimpath for tests

### PR DESCRIPTION
**Pull request by Mic92**
###### Description of changes

Not yet widely tested but inspired by https://github.com/NixOS/nixpkgs/pull/182647/files#r928199698

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/go-modules/generic/default.nix [hunk 1]
```diff
@@ -249,6 +249,8 @@ let
     doCheck = args.doCheck or true;
     checkPhase = args.checkPhase or ''
       runHook preCheck
+      # We do not set trimpath for tests, in case they reference test assets
+      export GOFLAGS=${lib.optionalString (!proxyVendor) "-mod=vendor"}
```
Comment (Mic92) [inline pkgs/development/go-modules/generic/default.nix:253]: Also overriding GOFLAGS might cause other regressions:

```suggestion
      export GOFLAGS=''${GOFLAGS//-trimpath/}
```

**Comment by zowoq** [comment]
With the number of rebuilds this should go to staging.

**Comment by zowoq** [comment]
Please also add this to `buildGoPackage`.

**Comment by Mic92** [comment]
> Please also add this to `buildGoPackage`.

done.

**Review by zowoq — APPROVED** [review]

**Comment by kira-bruneau** [comment]
I bisected the build failure in localtime to this commit:
```
localtime-unstable> checking for references to /build/ in /nix/store/ls326km7n977nj8hg1szaciiz01fa2gw-localtime-unstable-2022-02-20...
error: output '/nix/store/ls326km7n977nj8hg1szaciiz01fa2gw-localtime-unstable-2022-02-20' is not allowed to refer to the following paths:
         /nix/store/i0rdvlkvc5sz0dr87s379j0i4xmnwrp5-go-1.18.4
```

It looks like the problem is that localtime runs the `go build` in the `installPhase`, and relies on `-trimpath` being set then. We could add a workaround specifically for localtime, but would it make sense to restore the `-trimpath` after the `checkPhase`?

**Comment by zowoq** [comment]
> would it make sense to restore `-trimpath` after the `checkPhase`?

I don't think so. `localtime` was incorrectly packaged anyway, it should have already been using a separate buildPhase. Also shouldn't have been building and just ignoring the binaries produced by the `buildGoModule` default buildPhase.

**Pull request merged**